Refine query variables and clauses pages (#970)
graknlabs/docs: new_core_concepts/modules/ROOT/pages/typeql/query-variables-patterns.adoc

[,typeql]
----
define

>     entity person;      # 'person' is an entity type.

Author: Tab = 2 spaces

What does it mean to [[definition-match]]"*match*" a pattern?
A conjunctive pattern is "satisfied" if each variable can be mapped to some concept
so that every constraint in the pattern holds true in the database,
and every subpattern is also satisfied.

Each such mapping from variables to concepts is an [[definition-answer]]*answer* to the pattern.

> For the query above, each answer maps `$e` to an entity, and `$a` to an attribute owned by that type.

Reviewer: I might actually prefer to use entity *type* and attribute *type* here.
I just avoided it in the first section because it was very introductory.

Author: But it's incorrect here, innit? The query above maps entities to vars, not entity types to vars. There are `isa`s.

Reviewer: XD
Wow I didn't realise I inserted so many sections in between. I wrote that for the snippet way on top, just after the define. So this whole line has to be updated.

Author: Oh shoot, I checked that the person exists there and put a checkmark. Fixed

and every subpattern is also satisfied.

Each such mapping from variables to concepts is an [[definition-answer]]*answer* to the pattern.

> For the query above, each answer maps `$e` to an entity, and `$e` to a relation linking that entity.

Reviewer: Remove this line please. It was meant for the example above and no longer serves a purpose here.

Author: Okay =(

  ## $p is a `person`, who has the attribute $n
  $p isa person; $p has $n;

  ## $e is an 'employment', where $p plays the 'employee' role
  $e isa employment; $e links (employee: $p);
----
[NOTE]
====
Variables are also bound to concepts when they are inserted by an `isa` constraint.
See xref:{page-version}@new_core_concepts::typeql/schema-data.adoc[].
====

== A quick note on variables
Variable names are unique within a pipeline -
every usage of a variable with a certain name refers to the same variable.

In the following sections, the term [[definition-local-variable]]"local" is used to mean
a variable is bound in that pattern but not in its parent.
This contrasts with the usage in programming languages where it refers to the scope of a variable name.

> As we have seen, the scope of variable names in TypeQL is the xref:{page-version}@new_core_concepts::typeql/query-clauses.adoc[pipeline].

Reviewer: Maybe we don't need this sentence after all, since we only said it two sentences ago.
Else: 
```suggestion
As mentioned earlier, the scope of variable names in TypeQL is the xref:{page-version}@new_core_concepts::typeql/query-clauses.adoc[pipeline].
```

Author: Changed to "mentioned above". I got a little scratch of my eyes when I saw it, but didn't pay enough attention to fix

== Disjunctions, negations & optionals

We often want to write more complex patterns than pure conjunctions, which mean *all* the subpatterns must be true.

> This section describes the meaning of disjunctive, negated, and optional patterns.

Reviewer: Idk if I'm describing the meaning. I can describe the patterns themselves, or just straight out define the meaning. 
But I needed a word less strong than define, so went for discuss.

Author: You resulted in "discuss" ... "discuss", so I just went for a close word. I think that "straight out defining" is a form of describing, too.

=== Negations
A pattern is negated by wrapping it in a `not` block.

> An answer "satisfies" the negation if it cannot be (part of) an answer to the negated pattern.

Reviewer: awww i like iff. It's a much stronger term than if :(

Author: Hahahaha. "awww", indeed. It makes it not stronger. It's like there is a little hedgehog getting ready to sneeze.

Reviewer: Unless you mentally expand it to [https://en.wikipedia.org/wiki/iff]("if and only iff") as intended

Author: Let me put "if and only if" instead
